File: rishabh/387. First Unique Character In a String.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        ## Approach-1: brute force
        ## Time: O(n * n)
        ## Space: O(1)
        
        for i in range(len(s)):
            # False flag denotes there is no repitition
            flag = False                     
            for j in range(len(s)):
                if i != j:                          # skip element at the same index
                    if s[i] == s[j]: flag = True    # found a duplicate
            if not flag: return i

        return -1       # all characters repeat


    def firstUniqCharSets(self, s: str) -> int:    
        ## Approach-2: using sets
        ## use 2 sets, 1st to store every new element, 
        ##             2nd to store if an ele occurs again
        ## traverse the list again and return the char that occurs in 1st but not in 2nd 
        ## Time: O(n)
        ## Space: O(n)
        
        first, second = set(), set()

        for c in s:
            if c not in first: first.add(c)         # store every new element in 1st
            elif c in first: second.add(c)          # if an ele occurs again, store in 2nd

        # traverse the list again and return the char that occurs in 1st but not in 2nd     
        for idx,c in enumerate(s):
            if c in first and c not in second:
                return idx
        
        return -1
        
    def firstUniqChar3(self, s: str) -> int:
        ## Approach-3: Using Dictionary
        ## dict preserves order
        ## return the first key where val == 1
        ## Time: O(n)
        ## Space: O(n)
        count = {}      # {char: occurance}
        for c in s:
            count[c] = count.get(c, 0) + 1

        non_repeat_char = ''
        # print(count)

        for k,v in count.items():
            if v == 1:                  # pick the first key for which val is 1
                non_repeat_char = k
                break

        for idx,ele in enumerate(s):      # return above picked key's idx
            if ele == non_repeat_char:
                return idx

        return -1
    
    # short version of firstUniqChar3()
    def firstUniqChar4(self, s: str) -> int:
        count = {}      # {char: occurance}
        for c in s:
            count[c] = count.get(c, 0) + 1

        for idx in range(len(s)):
            if count[s[idx]] == 1: return idx
        
        return -1
    

File: rishabh/387. First Unique Character In a String.java
package rishabh;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;

@SuppressWarnings("unused")
class Solution {
    public static void main(String[] args) {
        System.out.println(firstUniqCharBruteForce("lleetcode"));
        System.out.println(firstUniqCharUsingSets("leetcode"));
        System.out.println(firstUniqCharUsingDictionary("lleettcode"));
    }

    // Approach-3: Using dictionary
    // Time: O(n)
    // Space: O(n)
    public static int firstUniqCharUsingDictionary(String s) {
        // if you want to maintain order
        // LinkedHashMap<Character, Integer> counter = new LinkedHashMap<>();
        HashMap<Character, Integer> counter = new HashMap<>();

        for (char c : s.toCharArray()) {
            counter.put(c, counter.getOrDefault(c, 0) + 1);
        }

        for (int i = 0; i < s.length(); i ++) {
            if (counter.get(s.charAt(i)) == 1){
                return i;
            }
        }
        return -1;
    }

    // Approach-2: using 2 hashsets
    // Time: O(n)
    // Space: O(n)
    public static int firstUniqCharUsingSets(String s) {
        HashSet<Character> first = new HashSet<>();
        HashSet<Character> second = new HashSet<>();

        for (char c : s.toCharArray()) {
            if (!first.contains(c)){
                first.add(c);
            } else {
                second.add(c);
            }
        }

        for (int i = 0; i < s.length(); i++) {
            if (!second.contains(s.charAt(i))){
                return i;
            }
        }
        return -1;
    }

    // Brute Force
    // Time: O(n * n)
    // Space: O(n)
    public static int firstUniqCharBruteForce(String s) {
        for (int i = 0 ; i < s.length(); i ++) {
            boolean duplicate = false;
            for (int j = 0; j < s.length(); j ++) {
                if ((i != j) && (s.charAt(i) == s.charAt(j))) {
                        duplicate = true;
                    }
                }
            if (!duplicate) {return i;}
            }
        return -1;
        }
}
